drivers/sdhc/Kconfig.brcm

```diff
@@ -21,14 +21,22 @@ config SDHC_BUFFER_ALIGNMENT
 
 choice BRCM_BCM2712_XFER_MODE
 	bool "Select BCM2712 SDHC (SD/MMC) data transfer mode"
-	default BRCM_BCM2712_XFER_SDMA
```

Comment:
Let's discuss this change in more depth. I think with the current implementation of ADMA2, there isn't any compelling reason to use it.

Here are the main differences I see compared to SDMA mode:
- ADMA2 may support 64-bit addresses, but SDMA might also support them. I haven't investigated this question thoroughly; I only know that the driver currently works only with 32-bit addresses.
- ADMA2 can transfer more data than the `block count` register can hold, but this functionality hasn't been implemented.
- **The main feature, in my opinion, is the ability to link descriptors and use non-linear physical memory chunks**. However, this hasn't been implemented (I even don't know if it is needed for Zephyr), and you are essentially wasting CPU time filling ADMA2 descriptors to function like SDMA.

Reply:
Here is the corrected version with improved clarity and grammar:

> Therefore, it supports only ADMA2 32-bit descriptors.

This statement is not true.

> The 64-bit descriptor is removed from the Simplified Specification because 64-bit system support will be modified by the Host Controller Specification Version 4.00.

This means that the 64-bit **descriptor** is not described in the **simplified** documentation. I suppose you don't have the full specification. Even in the simplified spec, there is a lot of information about the support of 64-bit addresses.

Finally, if you have time to open the spec version 4.00, you will find the following:

```
Support of 64-bit addressing is modified from Version 4.00. ADMA2 and ADMA3 64-bit descriptor length is modified to 128 bits considering byte enable alignment. 32-bit or 64-bit addressing mode is selected at initialization, which is determined by the OS. (Up to Version 3.00, 64-bit mode may be selected by the Host Driver at each operation by DMA Select in Host Control 1 register).
```

So, the descriptor format was changed; before v4, in the case of 64-bit addressing, we had 96 bits:
![Screenshot from 2024-07-05 09-10-32](https://github.com/xen-troops/zephyr/assets/27217224/c0bbbce6-0344-43c5-a6db-2af7c22950a9)

> Making transfers larger than the block count register can hold is, most probably, not a case for Zephyr (it's ~4G transfer :)).

First of all, you are making generic code. I don't understand how you calculated 4G. The block count register contains 16 bits according to spec v3, so you can have a maximum of 65535 blocks. The maximum possible block size is 2048 bytes according to the spec, so 65535 * 2048 bytes is approximately 128 Mbytes. Please explain how you arrived at 4G.

Finally, you don't know if you might have a case of a 1-byte block size, which is possible according to the SD Host Control specification.

Here is the corrected version with improved clarity and accuracy:

> The main difference between SDMA and ADMA2 is that SDMA has a limitation on uninterrupted transfer size, which is 512K now, while ADMA2 doesn't have it (taking into account ~4G max transfer size).

This statement is not true either.

Your ADMA2 implementation has similar limitations. Let's calculate: the maximum size of data in one descriptor is 2^16 bytes, and you have only 2^4 descriptors. So, in total, you will have 2^20 bytes, which is 1 Mbyte. To achieve this, you need to fill 16 descriptors. In the case of ADMA, you need to fill the address register two times. I'm not trying to say that you should avoid using ADMA2; this is just for your information.

To use ADMA2 for transferring 4 Gbytes, you need to have 2^16 descriptors, i.e., 0.5 Mbytes per descriptor array (in the case of 32-bit addresses). This is not a problem as per my understanding; it is just for clarification.

Reply:
> The 64-bit ADMA2 transfer is not supported by "SD Host Controller Simplified Specification Version 3.00" 

Please reread the comment above, and you will find that 64-bit mode is possible.

> it's as the fact there are no required control bits defined in relevant registers (Like Host Control1->DMA Select).
May be somebody has fulls version and need it - This person can easily create and send patch to extend this implementation.

You don't need a full spec; you see that spec 3.0 says the info is moved to spec 4.0. Have you tried to open spec 4.0 or later?

![Screenshot from 2024-07-10 08-49-38](https://github.com/xen-troops/zephyr/assets/27217224/355ab219-25b0-474d-927b-fc2777ff86af)

Some info from spec v3.0:
![Screenshot from 2024-07-10 08-45-18](https://github.com/xen-troops/zephyr/assets/27217224/6923d86a-a874-4dfe-9fd3-68fe5abb1f11)

There are many mentions of support for 64-bit mode in spec v3.0.

> For my use case 32-bit is enough right now and perfectly working.

Hmmm, it will be easy to get a 64-bit addresses when working under Xen. You don't know which pool of addresses will be allocated for the domain, and of course, there are versions of the RPi5 with more than 4GB of RAM.

If we talk more broadly, you are writing a common part of the driver that can be used not only by RPI5 boards, so it is possible that other boards will have addresses above 32 bits.

> Above you wrote a lot of word, but what do you propose exactly is probably hidden somewhere deep inside and I failed to see it.

As it was before, add support for 64-bit mode. **Of course, you could just open an issue for that.** I'm not insisting on fixing it right now; it would be enough to check the address and, in case of a 64-bit address, just redirect it to non-DMA mode. But here, you are just saying that it isn't supported, and your statement is really strange if we take into account the direct speech from the standard in my comments.

I think we need a referee here. @lorc FYI.

Reply:
I completely agree with you. It's agreed that 64-bit ADMA mode will not be implemented.